### PR TITLE
Socket's connect operations tracing

### PR DESCRIPTION
**Pull request by alnikola**
Socket's connect start, stop and cancelled events are logged as activities via EventSource.

Contributes to #37428

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System.Net.Sockets.csproj [hunk 1]
```diff
@@ -48,48 +49,29 @@
     <Compile Include="System\Net\Sockets\OverlappedAsyncResult.cs" />
     <Compile Include="System\Net\Sockets\ReceiveMessageOverlappedAsyncResult.cs" />
     <Compile Include="System\Net\Sockets\UnixDomainSocketEndPoint.cs" />
-    <Compile Include="$(CommonPath)System\IO\StreamHelpers.CopyValidation.cs"
-             Link="Common\System\IO\StreamHelpers.CopyValidation.cs" />
+    <Compile Include="$(CommonPath)System\IO\StreamHelpers.CopyValidation.cs" Link="Common\System\IO\StreamHelpers.CopyValidation.cs" />
```
Comment (ManickaP) [inline src/libraries/System.Net.Sockets/src/System.Net.Sockets.csproj:52]: Could you revert the unnecessary csproj changes? I.e. keep the element split over 2 lines.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 2]
```diff
@@ -2424,6 +2426,9 @@ public void EndConnect(IAsyncResult asyncResult)
             Exception? ex = castedAsyncResult.Result as Exception;
             if (ex != null || (SocketError)castedAsyncResult.ErrorCode != SocketError.Success)
             {
+                SocketsTelemetry.Log.ConnectFailed();
+                SocketsTelemetry.Log.ConnectStop();
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2430]: The general pattern is to guard call sites with an IsEnabled check:
```C#
if (SocketsTelemetry.Log.IsEnabled())
{
    SocketsTelemetry.Log.ConnectFailed();
    SocketsTelemetry.Log.ConnectStop();
}
```
That not only helps to minimize overhead when tracing isn't enabled, it helps the linker to strip out all of these methods when the linker's been told to omit EventSource-based tracing from the linked app.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        public void ConnectFailed()
+        {
+            WriteEvent(eventId: 3);
+        }
+
+        [Event(4, Level = EventLevel.Warning)]
+        public void ConnectCancelled()
+        {
+            WriteEvent(eventId: 4);
+        }
```
Comment (scalablecory) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:58]: Should these check `IsEnabled()`, or is `WriteEvent` checking that itself and cheap when disabled?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        [NonEvent]
+        private unsafe void WriteSocketEvent(int eventId, string address)
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:61]: Why is this needed?  Why can't ConnectStart just call `WriteEvent(int, string)`?
https://github.com/dotnet/runtime/blob/40b2e886a3e978928e8d1571df0852abba0cfb9d/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L869

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+{
+    [EventSource(Name = "System.Net.Sockets")]
+    internal sealed class SocketsTelemetry : EventSource
+    {
+        public static readonly SocketsTelemetry Log = new SocketsTelemetry();
+
+        [NonEvent]
+        public void ConnectStart(Internals.SocketAddress address)
+        {
+            if (IsEnabled())
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:17]: These IsEnabled checks need to check for the proper event level.  Otherwise, for example, in this case if the EventSource is enabled but not at the Informational level, we're still going to end up paying for the ToString operation.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Tracing;
+
+namespace System.Net.Sockets
+{
+    [EventSource(Name = "System.Net.Sockets")]
+    internal sealed class SocketsTelemetry : EventSource
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: We now have two EventSources in this library.  Is that really what we want?  I'm not clear on why the decision was made to add a new one.  I really don't want to be in a position of having to carry both forward indefinitely.
cc: @noahfalk, @brianrob, @davidsh

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        [NonEvent]
+        public void ConnectStart(EndPoint address)
+        {
+            if (IsEnabled())
+            {
+                var addressString = address.ToString();
+                if (addressString != null)
+                {
+                    ConnectStart(addressString);
+                }
```
Comment (scalablecory) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:32]: This looks _super_ corner-casey, but if I had counters enabled I would expect an entry for every connect -- instead of not logging on null, can we log e.g. `""` or `"<null>"`?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 8]
```diff
@@ -4338,6 +4345,9 @@ private void DoConnect(EndPoint endPointSnapshot, Internals.SocketAddress socket
             // Throw an appropriate SocketException if the native call fails.
             if (errorCode != SocketError.Success)
             {
+                SocketsTelemetry.Log.ConnectFailed();
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:4348]: It seems like every ConnectFailed is followed immediately by a ConnectStop.  Can't the former just do the latter and save the boilerplate/overhead at the call site?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 9]
```diff
@@ -720,12 +732,17 @@ internal void FinishOperationSyncSuccess(int bytesTransferred, SocketFlags flags
...
+                        SocketsTelemetry.Log.ConnectStop();
+
                         // Mark socket connected.
                         _currentSocket!.SetToConnected();
                         _connectSocket = _currentSocket;
                     }
                     else
                     {
+                        SocketsTelemetry.Log.ConnectFailed();
+                        SocketsTelemetry.Log.ConnectStop();
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:744]: Why have the Stop in both branches rather than just having it once after the else?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 5 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:17]: Should we attempt to avoid the ToString here entirely, or is the overhead considered acceptable? If the latter, similar logic in #38409 could be simplified.

**Review by scalablecory — COMMENTED** [review]
We have an established pattern for logging elsewhere of checking `Enabled` before calling the logging method, rather than checking it inside the method.

I prefer this new pattern's readability, but it looks like in some cases it will add some extra overhead when tracing is disabled. These are logged on a relatively expensive connect operation so overhead here will get swallowed, but still pointing out in case we use this pattern in less expensive areas.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 5 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: EventLevel check added.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs (hunk 2 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: Done

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs (hunk 8 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: Calls merged.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs (hunk 9 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs]: Initially, I aimed to call Stop as soon as it possible. However, now Failed calls Stop, so there is only one call on this line.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 3 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: IsEnabled checks now guard telemetry calls.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 7 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: Done

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 4 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: You are right, I somehow overlooked existing overloads. Fixed.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 5 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: This is called once per connection, so it seems acceptable. Additionally, I checked ToString() implementation and it looks quite efficient.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 6 above)
Comment (brianrob) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: Generally, you have one `EventSource` per area of functionality.  There's no hard and fast rule, but unless there's a reason to have multiple `EventSources` I'd recommend just having one.  It also makes it easier to consume the events.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        public void ConnectStop()
+        {
+            WriteEvent(eventId: 2);
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
+        {
+            WriteEvent(eventId: 3);
+            ConnectStop();
```
Comment (brianrob) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: This is a bit non-standard.  Generally if you want to trigger two events, you have a `[NonEvent]` method that calls both of them.  Generally speaking, `[Event]` methods should only log the event that they represent.

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: Can it lead to invalid log entries? It was done per above request to avoid unnecessary call to ConnectStop after ConnectFailed. I could have introduced a new private method like ConnectStopInternal that will be called from both of ConnectStop and ConnectFailed, but it looked redudant.

Comment (brianrob) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: I can't think of any reason why this would cause corruption or invalid log entries, but it is certainly non-standard.  If I understood the comment above, the goal was to not have two calls to it, one in each branch, and rather just call it afterwards unconditionally (for cleanliness).  I would recommend doing something like that rather than this to avoid chaining these events together.  The more you follow the standard `EventSource` rules, the better the chance things don't break in the future.  I can't point to anything that I think would break now though.

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: Extracted connection stop logging into a new ConnectionStopInternal method to be on the safe side.

**Comment by alnikola** [comment]
FYI. There is a typo in the [last commit](https://github.com/dotnet/runtime/pull/38620/commits/c693fe77e2780299b48acdb9e0057269b0791699) name. It's intended to be ConnectStopInternal.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: @alnikola, this doesn't need to block your PR, but I do want us to resolve this for .NET 5. Can you please make sure there's an issue for it?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
+        {
+            WriteEvent(eventId: 3);
+            ConnectStopInternal();
+        }
+
+        [Event(4, Level = EventLevel.Warning)]
+        public void ConnectCancelled()
+        {
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:53]: Cancelled => Canceled

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        public void ConnectStop()
+        {
+            ConnectStopInternal();
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
+        {
+            WriteEvent(eventId: 3);
+            ConnectStopInternal();
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: Maybe part of the strangeness here is: why do we need both events? Why not have a status be part of the stop, rather than logging one event for success and two for exception or cancellation. Also, the exception details don't need to be logged?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 13]
```diff
@@ -567,6 +571,12 @@ internal void CancelConnectAsync()
                 }
                 else
                 {
+                    if (SocketsTelemetry.IsEnabled(EventLevel.Warning))
+                    {
+                        SocketsTelemetry.Log.ConnectCancelled();
+                        SocketsTelemetry.Log.ConnectStop();
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:577]: Seems like this should be on the same plan as ConnectFailed, with just one log at the call site.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 6 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: [Issue already exists](https://github.com/dotnet/runtime/issues/37655), but it's outside .NET 5.0 scope.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 12 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: I assume that having status as part of the stop event can become an issue for Geneva or similar tools queries because it will require event attributes extraction whereas currently they can relate on event ID only. Additionally, this pattern has been already implemented in HttpClient telemetry (RequestStart/Stop/Aborted).

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: @brianrob, can you comment on best practices here?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: That is related but different.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 11 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:53]: Fixed

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: (and if we really want to use that same issue, the description and the milestone should be updated)

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: I created a dedicated issue #38754 for this, but it needs to be triaged to see if it can fit in .NET 5.0 scope.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: Thanks

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 12 above)
Comment (brianrob) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: Do you mean that Geneva would need to look at the fields in the event?  I would not expect that to be an issue, and from a perf perspective, it's cheaper to emit the status as a flag in the event than to emit a whole new event.  A reason that one might want to have a separate event for the failure is if the failure occurs at a different call stack than the stop, and that stack is useful for diagnostics.

My recommendation, unless you know that there is an issue with Geneva handling fields (I'm not aware of one), would be that you can put it as a flag on the stop event since the failed and stop events here will have the same stack.  Do make sure to keep the start and stop events so that it's easy to see how long the connection takes, and things work nicely with activity tracking.

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: But what about different EventLevels? Stop/Cancelled/Failed must be logged with different severity. How can it be achieved if we would have the only ConnectStop method with 3 status flags (OK, Cancelled, Failed)?

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: > Stop/Cancelled/Failed must be logged with different severity

Can you elaborate?  Do you mean you expect Canceled/Faulted to be logged even when Stop isn't?

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: Yes, that's correct, Stop is just Informational whereas Cancelled and Failed are of higher severity. It's done the same way as in [HttpTelemetry](https://github.com/dotnet/runtime/blob/87919a0462bc2ebd5c256e5ad7649ff98c858e2c/src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs#L51).

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 6 above)
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:10]: I added some thoughts to #38754 and may have more once I get a chance to look into this more

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 12 above)
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: IMO it is unlikely we'd get complaints if handling a failed socket connection consumed 1us to send an extra telemetry event. 

Given that start/stop are likely to be much higher volume events than errors I'd imagine customers having the Error level telemetry on all the time whereas start/stop would only get enabled when they needed more detailed performance tracing. That is a good reason to split out error as a separate 3rd event. Correlating the error event back to the stop event does add a little bit of complexity, but it is roughly the same as correlating the stop and start. If our tools have trouble doing this we should probably work to improve our tools rather than duplicate the error data into the stop event.

Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: However it is odd to make a single EventSource event API log two events. You might want to define a [NonEvent] API called something like ConnectFailedAndStop() or do two separate calls to the EventSource

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:48]: Could it cause some actual problems? They were combined to avoid redundant code.

**Review by noahfalk — COMMENTED** [review]
I made a few suggestions inline but the high order bit is probably to resolve whether a new EventSource should exist at all in #38754.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        [Event(2, Level = EventLevel.Informational)]
+        public void ConnectStop()
+        {
+            ConnectStopInternal();
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:45]: There is `SocketError` or `Exception`, but we don't log the details here because the goal is to provide telemetry for operation tracing/timings not debugging. Please, see how it has been already done in [HttpTelemetry](https://github.com/davidsh/runtime/blob/e5274e41fe13a8cfeeb4e87e39a1452cf8968f8c/src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs#L51).

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 15]
```diff
@@ -566,6 +570,11 @@ internal void CancelConnectAsync()
                 }
                 else
                 {
+                    if (SocketsTelemetry.IsEnabled(EventLevel.Warning))
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:573]: It's done here for consistency with all other cases.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 16]
```diff
@@ -553,6 +554,9 @@ internal void StartOperationConnect(MultipleConnectAsync? multipleConnect, bool
             _multipleConnect = multipleConnect;
             _connectSocket = null;
             _userSocket = userSocket;
+
+            // Log only the actual connect operation to a remote endpoint.
+            if (SocketsTelemetry.IsEnabled(EventLevel.Informational) && multipleConnect == null) SocketsTelemetry.Log.ConnectStart(_socketAddress!);
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs]: Done

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:42]: If the goal is solely to do timing analysis and not facilitate failure analysis then do we need a connect failed event at all? I assumed the purpose of this event having a different level is so that we can do failure analysis even while not doing timing analysis.

I probably should have asked the same question about the HttpTelemetry, I missed it there and caught it here : )

[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:42]: It's not only about timing, but also about number or rate of Successes/Failures, this is why we need `Failed`.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:42]: Exactly. This is my same question/concern with the design. I don't understand in what circumstance I care to differentiate why a connect stopped but not have more details on the failures.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 12 above)
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: I'm not aware it will cause any specific problems, just atypical usage that doesn't match suggested patterns others may be familiar with when they try to read your code.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:42]: Knowing the failure rate is quite useful for live service health monitoring/incident analysis because it's not uncommon for health to start degrading gradually so by observing failure rate you can catch it.
Having that said, I'm not against adding failure details (it's also useful), but we will need also to update implemented code (like HttpTelemetry).

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:42]: Failure rate is fine, but it should be part of the stop event, no? If we then add a second event for failure, it should include details.

[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:42]: [As discussed above](https://github.com/dotnet/runtime/pull/38620#discussion_r453504085), having a separate Failed event makes sense in this case taking into account that Stop/Fail have different levels.

[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: SocketError and Exception (if any) are logged now.

[the remark this answers is not in the document]
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: The only reason I see for Failure() to have a different level is if it plays a role in failure analysis. I recommend that is what we should do (for http as well), but if we didn't define the scenarios that way then the dominos would start falling:
Failure has no reason to be error -> Failure moves to be Info level -> Failure has no reason to be distinct event from stop -> failure event is removed and stop gains a pass/fail flag

[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: Failure **is** needed for failure analysis here and in HttpTelemetry as well.

[the remark this answers is not in the document]
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: I suspect we may be using the terminology differently and I'm happy to explain further if it would be helpful, but after you added the Exception/status code to the ConnectFailed() event my concerns on this are resolved. Thanks!

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: Let's say I enable start/stop events and the failed event.  How do I correlate an exact failed with an exact start/stop?  What data is included in all of these events that makes that feasible?

[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: Yes, EventSource infra automatically creates an activity for each start/stop pair and all nested events gets linked to parent ones (multi-level nesting is supported). Each activity has ID that is logged and used for correlation.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        [NonEvent]
+        private void ConnectStopInternal()
+        {
+            if (IsEnabled(EventLevel.Informational, EventKeywords.All))
+            {
+                WriteEvent(eventId: 2);
+            }
+        }
+    }
+}
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:71]: Please add at least basic tests.

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:71]: I could add them, but they were considered brittle and [it was agreed](https://github.com/dotnet/runtime/pull/37619) that telemetry logic is tested manually, thus other completed telemetry PRs do not have them.

>No automated tests were added at this time. The current discussions with others (reverse proxy
team) is that automated tests are brittle in CI. But manual tests will be run.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:71]: I disagree with that decision.  What's brittle is shipping important code with zero coverage of future changes breaking it.

We have some basic logging tests in Sockets, and I can't remember them every failing.
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Sockets/tests/FunctionalTests/LoggingTest.cs

It should not be hard to write a basic test that validates ConnectStart/Stop events are output.

Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:71]: Ok, I will add them.

**Comment by noahfalk** [comment]
> I prefer this new pattern's readability, but it looks like in some cases it will add some extra overhead when tracing is disabled

Assuming the if check is at the top of the EventSource method and you do no argument transformation beforehand the overhead should be call-setup/return. This varies based on arguments and machine, but I usually consider it in the 1ns order of magnitude. If performance on that scale matters for a given scenario then certainly we should benchmark and go from there.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 18]
```diff
@@ -2071,6 +2072,8 @@ private bool CanUseConnectEx(EndPoint remoteEP)
 
         internal IAsyncResult UnsafeBeginConnect(EndPoint remoteEP, AsyncCallback? callback, object? state, bool flowContext = false)
         {
+            if (SocketsTelemetry.Log.IsEnabled()) SocketsTelemetry.Log.ConnectStart(remoteEP);
```
Comment (noahfalk) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2075]: Style-wise I've always found it preferable to eliminate these IsEnabled() calls entirely and rely on the checks that occur inside the EventSource, but I think the mono-linker will fully remove this pattern so its purely style and tiny perf at this point. Style is best decided by the folks who have to work on the code rather than me : )

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 19]
```diff
@@ -2320,6 +2323,8 @@ public void EndConnect(IAsyncResult asyncResult)
             Exception? ex = castedAsyncResult.Result as Exception;
             if (ex != null || (SocketError)castedAsyncResult.ErrorCode != SocketError.Success)
             {
+                if (SocketsTelemetry.Log.IsEnabled()) SocketsTelemetry.Log.ConnectFailed((SocketError)castedAsyncResult.ErrorCode, ex);
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2326]: We want to measure timings as close to the wire as possible. It was requested on the issue #827

>The intent is to get these timings as low down in the stack and close as possible to the point that actual bytes are sent and received over the network so that accurate pictures can be developed

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 20]
```diff
@@ -4126,6 +4133,8 @@ private void DoConnect(EndPoint endPointSnapshot, Internals.SocketAddress socket
             // Throw an appropriate SocketException if the native call fails.
             if (errorCode != SocketError.Success)
             {
+                if (SocketsTelemetry.Log.IsEnabled()) SocketsTelemetry.Log.ConnectFailed(errorCode, null);
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:4136]: Same as above.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 21]
```diff
@@ -4134,6 +4143,8 @@ private void DoConnect(EndPoint endPointSnapshot, Internals.SocketAddress socket
                 throw socketException;
             }
 
+            if (SocketsTelemetry.Log.IsEnabled()) SocketsTelemetry.Log.ConnectStop();
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:4148]: Good point! Fixed by adding `ConnectStart` to `DoConnect`.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 17 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:69]: Done.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 12 above)
Comment (alnikola) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: Thanks for the idea! I changed it to ConnectFailedAndStop and ConnectCanceledAndStop.

**Review by ManickaP — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (noahfalk) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs]: If you want to see an example of the Activity tracing in action, there is sample code and some output included as the repro of this bug I filed last night while playing with it more: https://github.com/dotnet/runtime/issues/39353
